### PR TITLE
Fuse JAX Bidirectional LSTM into a single cuDNN call

### PR DESCRIPTION
**Pull request by MarcosAsh**
## Description

`Bidirectional(LSTM(...))` on the JAX backend issues two separate cuDNN LSTM calls today, one for the forward layer and one for the backward layer. cuDNN already supports a single bidirectional pass that handles both directions in one kernel launch, exposed via `jax.experimental.rnn.lstm(..., bidirectional=True)`.

I added a backend helper that invokes that fused kernel and a dispatch hook on `Bidirectional.call()` that routes to it when both children are plain LSTMs with no mask, no dropout, no stateful, matching units / activations / dtype, and `use_bias=True`. Anything outside that window continues through the existing two-call path.

GRU is unchanged because `jax.experimental.rnn` does not expose a cuDNN GRU binding.

### Benchmark

Inference on a Colab A100, JAX backend, `use_cudnn="auto"`, `return_sequences=True`:

| shape (b x s x in x h) | unfused  | fused    | speedup |
| ---------------------- | -------- | -------- | ------- |
| 32 x 128 x 128 x 128   | 3.95 ms  | 3.54 ms  | 1.12x   |
| 64 x 256 x 256 x 256   | 12.16 ms | 10.68 ms | 1.14x   |
| 64 x 512 x 256 x 256   | 22.04 ms | 21.06 ms | 1.05x   |
| 32 x 1024 x 512 x 512  | 53.08 ms | 48.12 ms | 1.10x   |

Modest but consistent. The dispatch is conservative; non-eligible configs cost nothing.

## Contributor Agreement

- [x] I am a human, and not a bot.
- [x] I will be responsible for responding to review comments in a timely manner.
- [x] I will work with the maintainers to push this PR forward until submission.

### CODE REVIEW

**Comment by hertschuh** [comment]
Sorry for the breakage in master, please rebase.

**Comment by MarcosAsh** [comment]
Hi, just rebased! Many thanks.

File: keras/src/layers/rnn/bidirectional.py [hunk 1]
```diff
@@ -200,6 +201,12 @@ def call(
         mask=None,
         training=None,
     ):
+        if self._can_use_jax_cudnn_fused(mask):
+            try:
+                return self._call_jax_cudnn_fused(sequences, initial_state)
+            except NotImplementedError:
+                pass
```
Comment (hertschuh) [inline keras/src/layers/rnn/bidirectional.py]: Can we make this backend independent by following the same pattern as the other operations (gru and lstm)? That is, each backend can implement `bidirectional_lstm` or throw `NotImplementedError`.

This will allow the addition of the same optimization for other backends if we ever wanted to.

**Review by hertschuh — CHANGES_REQUESTED** [review]
Thanks for your work on this!

**Comment by MarcosAsh** [comment]
Hi, refactored to dispatch via `backend.bidirectional_lstm` matching the `lstm`/`gru` pattern. Thanks!

File: keras/src/layers/rnn/bidirectional.py [hunk 2]
```diff
@@ -258,6 +265,112 @@ def call(
...
+
+        sequences = ops.convert_to_tensor(sequences)
+        batch_size = ops.shape(sequences)[0]
+
+        if initial_state is None:
+            zeros = ops.zeros((batch_size, units), dtype=sequences.dtype)
+            fwd_h0, fwd_c0, bwd_h0, bwd_c0 = zeros, zeros, zeros, zeros
+        else:
+            if len(initial_state) != 4:
+                raise NotImplementedError
```
Comment (hertschuh) [inline keras/src/layers/rnn/bidirectional.py]: Can you move this test to `_can_attempt_fused_lstm`? You'll have to pass the `initial_state` to it.

File: keras/src/layers/rnn/bidirectional.py [hunk 3]
```diff
@@ -258,6 +265,112 @@ def call(
...
+        # Layer-level preconditions for dispatching to
+        # `backend.bidirectional_lstm`. Backends that don't support a
+        # fused path (or don't have the runtime resources for one, e.g.
+        # no GPU) raise `NotImplementedError` from the call itself, and
+        # the caller falls back to the two-layer path.
+        if mask is not None or self.stateful:
+            return False
+
+        from keras.src.layers.rnn.lstm import LSTM
+
```
Comment (hertschuh) [inline keras/src/layers/rnn/bidirectional.py]: Move this import to the top (unless you can't)

**Review by hertschuh — CHANGES_REQUESTED** [review]
Thanks for adding this!

**Comment by MarcosAsh** [comment]
Hi, hoisted the `initial_state` arity check into `_can_attempt_fused_lstm` and moved the `LSTM` import up. Thanks!

**Review by hertschuh — DISMISSED** [review]
Thanks!

**Review by hertschuh — CHANGES_REQUESTED** [review]
Please look at the tests, the numerics don't match on GPU (i.e. when this is actually being used)

**Comment by MarcosAsh** [comment]
Hi, thanks for the check! The flat weight blob was packed as [fwd_W, fwd_b, bwd_W, bwd_b], but jax.experimental.rnn expects all W's across both directions first, then all b's. Wrong layout made backward weights and forward biases decode as garbage. Pushed a fix.

**Review by hertschuh — APPROVED** [review]
Thanks!
